### PR TITLE
subapps can handle errors and app.route works fine

### LINKED ISSUE #12: Subapp error handler is not respected
Reproduction:

```js
import { App } from 'https://denopkg.com/deno-libs/tinyhttp@master/mod.ts'
import { makeFetch } from 'https://deno.land/x/superfetch@1.0.4/mod.ts'

const app = new App({
  onError: (err, req) =>
  new Response(`Ouch, ${err} hurt me on ${req?.url} page.`, {
    status: 500,
  })
})

const subApp = new App({
  onError: (err, req) =>
  new Response(`Handling ${err} from child on ${req?.url} page.`, {
    status: 500,
  })
})

subApp.get('/route', async (req, res, next) => await next('you'))

app.use('/subapp', subApp)

const server = app.handler
const fetch = makeFetch(server)

const res = await fetch('/subapp/route')
res.expectStatus(500).expectBody('Handling you from child on /subapp/route page.')
```

### PR DESCRIPTION
**Pull request by aarontravass**
fixes #12 and #13 

I skipped 3 due to a weird dilemma.

review only after #15 is merged since I need to merge master branch changes.

### CODE REVIEW

File: app.ts [hunk 1]
```diff
@@ -56,12 +58,16 @@ export class App<
   notFound: Handler<Req, Res>
   attach: (req: Req, res: Res, next: NextFunction) => void
 
+  // this symbol tells if a custom error handler has been set
+  private readonly [hasSetCustomErrorHandler]: boolean
```
Comment (talentlessguy) [inline app.ts:62]: use ESNext private properties instead of TS ones

File: app.ts [hunk 2]
```diff
@@ -244,41 +246,43 @@ export class App<
...
 
     await loop()
 
-    if (err) throw err
+    if (err instanceof Response) throw err
+    else if (err) {
+      if (this[hasSetCustomErrorHandler]) throw await this.onError(err, req)
+      else throw err
+    }
+    throw new Response(res._body, res._init)
```
Comment (talentlessguy) [inline app.ts:283]: why do you throw a response?

File: url.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  return path
+}
+
+/**
+ * This function joins a list of paths with / (forward slash)
+ * @param {boolean} padLeft - pads the final path with leading slash if it does not exist
+ * @param {boolean} padRight - pads the final path with trailling slash if it does not exist
+ * @param {string[]} [path] - array of paths.
+ */
+const pathJoin = (padLeft: boolean, padRight: boolean, ...path: string[]) => {
```
Comment (talentlessguy) [inline url.ts:16]: why not use `new URL` instead? or std/path?

File: url.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
+const stripSlash = (path: string) => {
```
Comment (talentlessguy) [inline url.ts:1]: this function is not used anywhere, why create it?

**Review by talentlessguy — CHANGES_REQUESTED** [review]

File: app.ts [hunk 5]
```diff
@@ -244,41 +246,43 @@ export class App<
...
-
+        if (!req.path) req.path = req._urlObject.pathname
         if (this.settings?.enableReqRoute) {
           req.route = getRouteFromApp(this.middleware as any, handler as any)
         }
         await applyHandler<Req, Res>(handler)(req, res, next)
       }
 
-    let idx = 0, err: unknown | undefined
+    let idx = 0, err: unknown | undefined = null
```
Comment (talentlessguy) [inline app.ts:264]: why set to null?

File: app.ts [hunk 6]
```diff
@@ -244,41 +246,43 @@ export class App<
...
     const next: NextFunction = async (error) => {
-      if (error) err = error
+      if (error) {
```
Comment (talentlessguy) [inline app.ts:266]: why wrap in `{}`?

**Review by talentlessguy — CHANGES_REQUESTED** [review]

File: url.ts (hunk 3 above)
Comment (aarontravass) [reply url.ts:16]: `new URL` expects a standard URL
```ts
> new URL('/route')
Uncaught TypeError: Invalid URL: '/route'
    at getSerialization (ext:deno_url/00_url.js:88:11)
    at new URL (ext:deno_url/00_url.js:383:27)
    at <anonymous>:2:1
```

`path.join` joins with a platform specific delimiter. In linux and macos, this is `/` but in windows it is `\`. See https://stackoverflow.com/a/45575007

File: app.ts (hunk 2 above)
Comment (aarontravass) [reply app.ts:281]: I need to persist the response. If I don't throw, the response gets lost in the function stack during recursion.

File: app.ts (hunk 1 above)
Comment (aarontravass) [reply app.ts]: I changed it. But symbols are private I believe and I can't add a `#` in front since it's a runtime property.

**Comment by aarontravass** [comment]
one test is failing. specifically, 
```ts

Request properties > should have default HTTP Request properties => https://deno.land/x/tincan@1.0.1/src/runner.ts:166:10
error: AssertionError: Values are not equal: Expected to have body [object Object], got [object Object]


    [Diff] Actual / Expected


    {
-     url: "http://localhost:31437/",
+     url: "http://localhost:8080/",
    }

  throw new AssertionError(message);
        ^
    at assertEquals (https://deno.land/std@0.182.0/testing/asserts.ts:189:9)
    at Object.expectBody (https://deno.land/x/superfetch@1.0.5/mod.ts:149:7)
    at ItNode.fn (file:///home/runner/work/tinyhttp/tinyhttp/tests/core/request.test.ts:12:9)
    at eventLoopTick (ext:core/01_core.js:166:11)
    at async wrappedFn (https://deno.land/x/tincan@1.0.1/src/runner.ts:141:9)

Request properties > Network extensions > req.ip & req.ips is being parsed properly => https://deno.land/x/tincan@1.0.1/src/runner.ts:166:10
```
This pr https://github.com/deno-libs/superfetch/pull/3 should fix it since I can access the port and then verify the URL

**Comment by aarontravass** [comment]
Also, can you cancel hanging ci 
https://github.com/deno-libs/tinyhttp/actions/runs/5281150802
https://github.com/deno-libs/tinyhttp/actions/runs/5281095851

**Comment by talentlessguy** [comment]
@aarontravass I think instead of req.url we can test some other built-in property like method

File: url.ts (hunk 3 above)
Comment (talentlessguy) [reply url.ts:7]: @aarontravass you can use `std/path/posix`

File: app.ts [hunk 7]
```diff
@@ -244,17 +248,15 @@ export class App<
...
       (mw: Middleware<Req, Res>) =>
       async (req: Req, res: Res, next: NextFunction) => {
         const { handler, type, pattern } = mw
-
         const params =
           type === 'route' && pattern?.exec(req.url)?.pathname.groups || {}
 
         req.params = params as Record<string, string>
-
+        if (!req.path) req.path = req._urlObject.pathname
```
Comment (talentlessguy) [inline app.ts:259]: req path is already set here

https://github.com/deno-libs/tinyhttp/blob/b48d0a06a924a7b01460fe91f21abd99acabf73f/extend.ts#L57

**Review by talentlessguy — CHANGES_REQUESTED** [review]

File: url.ts (hunk 3 above)
Comment (talentlessguy) [reply url.ts:7]: tbh having a custom function is ok since the one in std is meant for fs

**Pull request merged**